File: LeetCode/Trapping_rain_water.cpp
class Solution {
public:
    void printVector(vector<int> &v){
        cout<<"the input size is "<<v.size()<<endl;
        for(vector<int>::iterator it = v.begin(); it != v.end(); it++){
            cout<<*it<<" ";
        }
        cout<<"vector Ends Here"<<endl;
    }
    int trap(vector<int>& height) {
        int inputSize = height.size();
        if(inputSize == 0){
            return 0;
        }
        
        vector<int> leftHeights(inputSize);
        vector<int> rightHeights(inputSize);
        
        leftHeights[0]= height[0];
        for(int iter = 1,n = height.size(); iter<n;iter++){
            if(leftHeights[iter-1] < height[iter]){
                leftHeights[iter] = height[iter];
            }
            else{
                leftHeights[iter] = leftHeights[iter-1];
            }
        }
        //printVector(leftHeights);
        rightHeights[height.size()-1] = height.back();
        for(int iter = height.size()-2; iter>=0;iter--){
            if(height[iter] > rightHeights[iter+1]){
                rightHeights[iter] = height[iter];
            }
            else{
                rightHeights[iter] = rightHeights[iter+1];
            }
        }
        //printVector(rightHeights);
        int answer = 0;
        for(int iter = 0;iter<inputSize;iter++){
            if(min(leftHeights[iter],rightHeights[iter]) - height[iter] > 0){
                answer = answer+ min(leftHeights[iter],rightHeights[iter]) - height[iter];
            }
        }
        return answer;
    }
};

File: LeetCode/trapping_rain_water.java
class Solution {
    public int trap(int[] height) {
         Stack<Integer> st = new Stack<Integer>();
        if(height.length == 0) return 0;
        for(int iter = 0; iter< height.length; iter++){
            if(st.isEmpty()){
                st.push(iter);
                //System.out.println("added value "+ iter+ " to stack");
            }else if(height[st.peek()] < height[iter]){
                st.push(iter);
                //System.out.println("added value "+ iter+ " to stack");
            }
        }
        
        int total = 0, right = height[height.length-1];
        for(int iter =height.length-1; iter>=0 ; iter--){
            if(iter == st.peek()){
                //System.out.println("iter is equal to stack top ");
                right = height[st.peek()];
                //System.out.println(" right is updated to "+ right);
                st.pop();
            }else{
                if(right < height[iter]){
                    right = height[iter];
                    //System.out.println("right is udpated to " + right);
                }
                if(Math.min(height[st.peek()] , right) > height[iter]){
                    total += (Math.min(height[st.peek()], right) - height[iter]);
                    //System.out.println(" total is updated to "+ total );
                    //System.out.println(" " + Math.min(height[st.peek()], right) + " - " + height[iter] + " = " + (Math.min(height[st.peek()], right) - height[iter]));
                }
                    
            }
        }
        return total;
    }
}